### ISSUE TITLE
Wrong self-convention lint wrongly emitted for `is_foo()` functions on builders

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust-clippy#8480]
### Summary

Builders using `&mut self` instead of `self` can cause the lint to trigger wrongly.

### Lint Name

wrong_self_convention

### Reproducer

I tried this code:

```rust
pub struct Builder {
    is_detailed: bool,
}

impl Builder {
    pub fn is_detailed(&mut self, value: bool) -> &mut Builder {
        self.is_detailed = value;
        self
    }
}
```

I saw this happen:

```
warning: methods called `is_*` usually take `self` by reference or no `self`
  --> src/codegen/signal_body.rs:31:24
   |
31 |     pub fn is_detailed(&mut self, value: bool) -> &mut Builder {
   |                        ^^^^^^^^^
   |
   = note: `#[warn(clippy::wrong_self_convention)]` on by default
   = help: consider choosing a less ambiguous name
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention
```

I expected to see this happen: No warning at all

### Version

```text
rustc 1.60.0-beta.1 (0a4f984a8 2022-02-22)
binary: rustc
commit-hash: 0a4f984a87c7ba6c74ec3e78442fec955a419e32
commit-date: 2022-02-22
host: x86_64-unknown-linux-gnu
release: 1.60.0-beta.1
LLVM version: 14.0.0
```
